### ISSUE TITLE
twisted.trial.loopback.loopbackAsync delivers data a little too quickly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3302]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3302|
|Type|defect|
|Created|2008-06-20 21:45:25Z|
|Branch|https://github.com/twisted/twisted/tree/async-loopback-3302|

`LoopbackTransport.write` can cause `dataReceived` to be called on the protocol belonging to it.  This is unrealistic and complicates application code unnecessarily.  The transport.write call should always be off the stack before a response to the write is delivered to the protocol.

<details><summary>Searchable metadata</summary>

```
trac-id__3302 3302
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_async_loopback_3302 branches/async-loopback-3302
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1213998325000000 1213998325000000
changetime__1214085906000000 1214085906000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [23953]) Branching to 'async-loopback-3302'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [23955]) add a unit test for not entering dataReceived while write is on the stack


refs [#3302](https://github.com/twisted/twisted/issues/3302)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Note also r23954

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

Flakes:
```
twisted/test/test_loopback.py:361: redefinition of unused 'reactor' from line 18
twisted/mail/test/test_pop3.py:10: 'string' imported but unused
```

Then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [23968]) remove unused import; update copyright date; twiddle whitespace

refs [#3302](https://github.com/twisted/twisted/issues/3302)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

also r23969

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [23970]) Merge async-loopback-3302

Author: exarkun
Reviewer: therve
Fixes: [#3302](https://github.com/twisted/twisted/issues/3302)

Change the implementation of twisted.protocols.loopback.lookbackAsync
so that it uses `IReactorTime.callLater` to "cut back" the stack when
resuming data delivery in response to a transport write.  This avoids
the case where a protocol's dataReceived method might be called when
it is in the middle of a call to its transport's write method.  Real
transports do not produce this behavior, since they typically must go
back into the event notification loop to learn about the written data.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
